cmd/db_open.go

```diff
@@ -0,0 +1,187 @@
+  tasks:
+    - name: Dump database credentials
+      template:
+        src: db_credentials.json.j2
+        dest: "{{ dest }}"
+        mode: '0600'
+      with_dict: "{{ wordpress_sites }}"
+      when: item.key == site
+`
+
+var dbCredentialsJsonJ2Template = `
+{
+    "web_user": "{{ web_user }}",
+    "ansible_host": "{{ ansible_host }}",
+    "ansible_port": {{ ansible_port | default(22) }},
+    "db_user": "{{ site_env.db_user }}",
+    "db_password": "{{ site_env.db_password }}",
+    "db_host": "{{ site_env.db_host }}",
+    "db_name": "{{ site_env.db_name }}",
+    "wp_env": "{{ site_env.wp_env }}"
+}
+`
+
+func NewDBOpenCommand(ui cli.Ui, trellis *trellis.Trellis, dbOpenerFactory *DBOpenerFactory) *DBOpenCommand {
+	c := &DBOpenCommand{UI: ui, Trellis: trellis, dbOpenerFactory: dbOpenerFactory}
+	c.init()
+	return c
+}
+
+type DBOpenCommand struct {
+	UI              cli.Ui
+	flags           *flag.FlagSet
+	app             string
+	Trellis         *trellis.Trellis
+	dbOpenerFactory *DBOpenerFactory
+}
+
+type DBCredentials struct {
+	SSHUser    string `json:"web_user"`
+	SSHHost    string `json:"ansible_host"`
+	SSHPort    int    `json:"ansible_port"`
+	DBUser     string `json:"db_user"`
+	DBPassword string `json:"db_password"`
+	DBHost     string `json:"db_host"`
+	DBName     string `json:"db_name"`
+	WPEnv      string `json:"wp_env"`
+}
+
+func (c *DBOpenCommand) init() {
+	c.flags = flag.NewFlagSet("", flag.ContinueOnError)
+	c.flags.Usage = func() { c.UI.Info(c.Help()) }
+	c.flags.StringVar(&c.app, "app", "", "Database client to be used; Supported: tableplus, sequel-pro")
+}
+
+func (c *DBOpenCommand) Run(args []string) int {
+	if err := c.Trellis.LoadProject(); err != nil {
+		c.UI.Error(err.Error())
+		return 1
+	}
+
+	if flagsParseErr := c.flags.Parse(args); flagsParseErr != nil {
+		return 1
+	}
+
+	args = c.flags.Args()
+
+	commandArgumentValidator := &CommandArgumentValidator{required: 1, optional: 1}
+	if commandArgumentErr := commandArgumentValidator.validate(args); commandArgumentErr != nil {
+		c.UI.Error(commandArgumentErr.Error())
+		c.UI.Output(c.Help())
+		return 1
+	}
+
+	environment := c.flags.Arg(0)
+	environmentErr := c.Trellis.ValidateEnvironment(environment)
+	if environmentErr != nil {
+		c.UI.Error(environmentErr.Error())
+		return 1
+	}
+
+	siteNameArg := c.flags.Arg(1)
+	siteName, siteNameErr := c.Trellis.FindSiteNameFromEnvironment(environment, siteNameArg)
+	if siteNameErr != nil {
+		c.UI.Error(siteNameErr.Error())
+		return 1
+	}
+
+	// Template JSON file for db credentials
+	dbCredentialsJson, dbCredentialsErr := ioutil.TempFile("", "*.json")
+	if dbCredentialsErr != nil {
+		c.UI.Error(fmt.Sprintf("Error createing temporary db credentials JSON file: %s", dbCredentialsErr))
+	}
+	defer deleteFile(dbCredentialsJson.Name())
+
+	// Template playbook files from package to Trellis
+	playbookPath := "dump_db_credentials.yml"
+	writeFile(playbookPath, templates.TrimSpace(dumpDbCredentialsYmlTemplate))
```

Comment:
Every usage of `writeFile` deletes the file after. Maybe we should just make this a more specific `writeTempFile` function which also handles the defer delete?

Reply:
How do we defer the `defer`?

If we put `defer` inside `writeTempFile`, it deletes teh files at the end of `writeTempFile` instead of the end of `func (c *DBOpenCommand) Run(args []string) int`

Reply:
Oh yes good call. We do this: https://github.com/roots/trellis-cli/blob/54aeb5d6fdf9df75afdb59c48fbca34d64fd2125/trellis/testing.go#L32-L38

Then it would be called as:

```go
defer writeFile(path, foo)()
```